code/core/write_sdl_ntuple.cc

```diff
@@ -199,114 +198,46 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
-
-    // Triplets from T5
-    unsigned int T3Idx1 = quintupletsInGPU.tripletIndices[2 * T5Idx];
-    unsigned int T3Idx2 = quintupletsInGPU.tripletIndices[2 * T5Idx + 1];
-
-    // LSs from T5
-    unsigned int LS1 = tripletsInGPU.segmentIndices[2 * T3Idx1];
-    unsigned int LS2 = tripletsInGPU.segmentIndices[2 * T3Idx1 + 1];
-    unsigned int LS3 = tripletsInGPU.segmentIndices[2 * T3Idx2];
-    unsigned int LS4 = tripletsInGPU.segmentIndices[2 * T3Idx2 + 1];
-
-    // MDs from T5
-    unsigned int pLS_md1 = segmentsInGPU.mdIndices[2 * pLSIdx];
-    unsigned int pLS_md2 = segmentsInGPU.mdIndices[2 * pLSIdx + 1];
-    unsigned int T5_md1 = segmentsInGPU.mdIndices[2 * LS1];
-    unsigned int T5_md2 = segmentsInGPU.mdIndices[2 * LS2];
-    unsigned int T5_md3 = segmentsInGPU.mdIndices[2 * LS3];
-    unsigned int T5_md4 = segmentsInGPU.mdIndices[2 * LS4];
-    unsigned int T5_md5 = segmentsInGPU.mdIndices[2 * LS4 + 1];
-
-    // Getting Hit indices
-    unsigned int pLS_hit1 = miniDoubletsInGPU.anchorHitIndices[pLS_md1];
-    unsigned int pLS_hit2 = miniDoubletsInGPU.outerHitIndices [pLS_md1];
-    unsigned int pLS_hit3 = miniDoubletsInGPU.anchorHitIndices[pLS_md2];
-    unsigned int pLS_hit4 = miniDoubletsInGPU.outerHitIndices [pLS_md2];
-    unsigned int T5_hit1  = miniDoubletsInGPU.anchorHitIndices[T5_md1];
-    unsigned int T5_hit2  = miniDoubletsInGPU.outerHitIndices [T5_md1];
-    unsigned int T5_hit3  = miniDoubletsInGPU.anchorHitIndices[T5_md2];
-    unsigned int T5_hit4  = miniDoubletsInGPU.outerHitIndices [T5_md2];
-    unsigned int T5_hit5  = miniDoubletsInGPU.anchorHitIndices[T5_md3];
-    unsigned int T5_hit6  = miniDoubletsInGPU.outerHitIndices [T5_md3];
-    unsigned int T5_hit7  = miniDoubletsInGPU.anchorHitIndices[T5_md4];
-    unsigned int T5_hit8  = miniDoubletsInGPU.outerHitIndices [T5_md4];
-    unsigned int T5_hit9  = miniDoubletsInGPU.anchorHitIndices[T5_md5];
-    unsigned int T5_hit10 = miniDoubletsInGPU.outerHitIndices [T5_md5];
 
     // pLS  1    2    3    4    5
     // **** o -- o -- o -- o -- o
     //      o -- o -- o
     //                o -- o -- o
-    unsigned int i1_idx = T5_hit1;
-    unsigned int i3_idx = T5_hit5;
-    unsigned int i5_idx = T5_hit9;
-
-    const float dr_in = sqrt(pow(hitsInGPU.xs[i3_idx] - hitsInGPU.xs[i1_idx], 2) + pow(hitsInGPU.ys[i3_idx] - hitsInGPU.ys[i1_idx], 2));
-    const float dr_out = sqrt(pow(hitsInGPU.xs[i5_idx] - hitsInGPU.xs[i3_idx], 2) + pow(hitsInGPU.ys[i5_idx] - hitsInGPU.ys[i3_idx], 2));
-    float betaIn_in   = __H2F(tripletsInGPU.betaIn[T3Idx1]);
-    float betaOut_in  = __H2F(tripletsInGPU.betaOut[T3Idx1]);
-    float betaIn_out  = __H2F(tripletsInGPU.betaIn[T3Idx2]);
-    float betaOut_out = __H2F(tripletsInGPU.betaOut[T3Idx2]);
+    unsigned int pT5 = trackCandidatesInGPU.directObjectIndices[idx];
+    std::vector<unsigned int> Hits = getHitsFrompT5(event, pT5);
+    unsigned int Hit_0 = Hits[0];
+    unsigned int Hit_4 = Hits[4];
+    unsigned int Hit_8 = Hits[8];
```

Comment:
This PR is very helpful in making the code more understandable but I feel some points may still be a bit cryptic. In this instance for example, why take the 9th hit of a pT5? Can there a comment next to each of the hits to explain what it corresponds to? There are more such instances below.

I add a suggestion below but I am not sure it is the correct one...
```suggestion
    unsigned int Hit_0 = Hits[0]; // First hit of pLS
    unsigned int Hit_4 = Hits[4]; // First hit of 1st OT LS
    unsigned int Hit_8 = Hits[8]; // First hit of 2nd OT LS
```

Reply:
The above "cartoon" comment was make this clear but clearly it was not sufficient!

Reply:
But `Hits` also include pLS hits, no? L320 of AccessHelper.cc seems to indicate so but maybe I am missing something?

Reply:
The cartoon is very helpful by the way, I would keep it, thanks!